### ISSUE TITLE
Rename counts (and counts64) to "sizes" and get the axis right

### ISSUE THREAD

**Issue by jpivarski** [scikit-hep/awkward#117]
(It's important that it's plural: `sizes`!)

This is a follow-on from https://github.com/scikit-hep/awkward-1.0/pull/115#issuecomment-586623849.

Given an array like

```python
[[[  2,   3,   5,   7,  11],
  [ 13,  17,  19,  23,  29],
  [ 31,  37,  41,  43,  47]],
 [[ 53,  59,  61,  67,  71],
  [ 73,  79,  83,  89,  97],
  [101, 103, 107, 109, 113]]]
```

   * `array.sizes(axis=0)` should return `[3, 3]` (as a NumpyArray)
   * `array.sizes(axis=1)` should return `[[5, 5, 5], [5, 5, 5]]` (as a RegularArray/ListArray/ListOffsetArray of NumpyArray)
   * `array.sizes(axis=2)` should be illegal.

I guess negative `axis` should count from the deepest possible, so `axis=-1` → `axis=1` and `axis=-2` → `axis=0`. Whatever this does, `flatten` should do the same, so this issue might supersede #51.

Let's consider some examples with branching, to figure out what they should do. An array like

```python
[[{"x": [], "y": [[]]},
  {"x": [1], "y": [[], [1]]},
  {"x": [2, 2], "y": [[], [1], [2, 2]]}],
 [],
 [{"x": [3, 3, 3], "y": [[], [1], [2, 2], [3, 3, 3]]},
  {"x": [4, 4, 4, 4], "y": [[], [1], [2, 2], [3, 3, 3], [4, 4, 4, 4]]}]]
```

has depth 3 along `"x"` and depth 4 along `"y"`.

   * `array.sizes(axis=0)` should return `[3, 0, 2]`.
   * `array.sizes(axis=1)` should return `[[{"x": 0, "y": 1}, {"x": 1, "y": 2}, {"x": 2, "y": 3}], [], [{"x": 3, "y": 4}, {"x": 4, "y": 5}]]`
   * `array.sizes(axis=2)` should be illegal because it fails for `"x"`. An exception occurs during the recursive descent; we don't have to check for it before descending.

Now for the negative `axis` values:

   * `array.sizes(axis=-1)` should return `[[{"x": 0, "y": [0]}, {"x": 1, "y": [0, 1]}, {"x": 2, "y": [0, 1, 2]}], [], [{"x": 3, "y": [0, 1, 2, 3]}, {"x": 4, "y": [0, 1, 2, 3, 4]}]]` because it represents the deepest `axis` along each branch.
   * `array.sizes(axis=-2)` should be illegal because it wants to be a non-record along `"x"` and a record along `"y"`.

So, negative `axis` values aren't just synonyms for non-negative `axis` values, as they are for rectilinear data. The reducer operations (PR #115) have a similar meaning, though reducers can be applied one level deeper than `sizes` and `flatten` can, so it's a little different.

I'm not sure whether @ianna or I will do this. We might need to figure it out together.

@nsmith-: you might have opinions about the usefulness of negative `axis` having a different meaning than positive `axis` minus depth.

**Comment by jpivarski** [comment]
For what it's worth, there's a new `branch_depth` method on all `Content` subclasses that reducers needed for their logic. As it turns out, `minmax_depth == purelist_depth` wasn't right for that case; I had to implement different behavior depending on whether any records or unions cause branches.

https://github.com/scikit-hep/awkward-1.0/blob/395def2a9481bbf07d30d1a37a7875ec619ceba5/src/libawkward/Content.cpp#L77-L97

It may be that my original `minmax_depth` and `purelist_depth` are irrelevant. If, for example, there are two RecordArrays in the tree and one splits the min/max into `(2, 3)` while the other unsplits them into `(4, 4)`, we still want to treat that as a branching depth, only allowing negative `axis` up to the first branch point. The new `branch_depth` method returns a boolean ("is the depth ever different?") and the minimum depth ("what's the largest possible `-axis`?").

Implementing `axis` for `sizes` might need to do something _similar_ to that. This is definitely something to work out in a Python-based study.

**Comment by nsmith-** [comment]
Another possible option is `ak.length` (or plural), one advantage is there is no `np.length` or `pd.length`, making it a bit more orthogonal.

**Comment by jpivarski** [comment]
True, and we should consider that. Internally in Awkward, however, `length` is always used as the number of elements in an array and `size` is used as the number of elements in each subarray of a RegularArray. The fact that we use this terminology internally isn't a strong argument for presenting a different terminology externally.

The lack of any antecedent in NumPy could be good or bad: it provides no guidance but it also doesn't misguide. The fact that `np.size(array, axis=X)` returns the number of elements at level `X` and we want `ak.sizes(array, axis=X)` to return the number of elements in each subarray at level `X + 1` is potentially an issue that another word would help with.

I've got to think about it.

(One thing, though, it definitely needs to be plural: `ak.lengths`.)

**Comment by ianna** [comment]
Oh, I remember that. It was implemented as described above, then I extended it to an extra depth to return a one dimensional NumpyArray as in `counts(axis)` from `study/flatten.py`

Either `lengths` or `sizes` is less confusing then `counts`.

Intuitively, I’d say `length` is constant while `size` isn’t, but it’s from Java/C++ world.

**Comment by jpivarski** [comment]
Some changes in definition, implemented in #152: (1) I shifted the meaning of `axis` by one, so that it's always describing lengths of the specified axis, rather than one deeper. So in my original example,

```python
[[[  2,   3,   5,   7,  11],
  [ 13,  17,  19,  23,  29],
  [ 31,  37,  41,  43,  47]],
 [[ 53,  59,  61,  67,  71],
  [ 73,  79,  83,  89,  97],
  [101, 103, 107, 109, 113]]]
```

   * `axis=0` now returns `2`, the scalar length of the array;
   * `axis=1` now returns `[3, 3]` (what the old `axis=0` would have returned)
   * `axis=2` now returns `[[5, 5, 5], [5, 5, 5]]` (what the old `axis=1` would have returned).
   * no `axis` values that correspond to an axis in the array (this one has only three: `0`, `1`, and `2`) are illegal, although `axis=0` is qualitatively different from the rest. (That's a recurring theme in these operations.)

Major change number (2) is that I've changed the name again, from "sizes" to "num". This will allow physicists to write very readable code, like

```python
ak.num(muons)     # implicit axis=1
```

and

```python
muons[ak.num(muons) >= 2]
```

There is no collision in the NumPy namespace with a function named "num", and it's far from any of NumPy's existing concepts (unlike "size").
